enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,478 @@
+
+`CCM` operator will provide initial manifests that allow to deploy `CCM` on the bootstrap machine with `bootkube.sh` [script](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template).
+
+At the initial stage of `CCM` installation installer creates a config map with `cloud.conf` key that contains configuration of `CCM` in `ini` format. The contents of `cloud.conf` are static and generated by the installer:
+
+Example for OpenStack:
+
+```txt
+[Global]
+secret-name = openstack-credentials
+secret-namespace = kube-system
+```
+
+Resources deployed here will be destroyed with the bootstrap stage cleanup.
+
+#### Operator resource management strategy
+
+Every cloud manages it's configuration differently, but shares a common overall structure.
+
+1. `cloud-controller-manager` and related resources are deployed under a single namespace.
+2. Each pod (`CCM`, `cloud-node-manager`, etc.) will be managed by a `Deployment`.
+3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.
+
+Operator will manage:
+
+- Static infrastructure resources creation, such as namespaces or service accounts will be delegated to the [CVO](https://github.com/openshift/cluster-version-operator).
+- Operator will populate, create and manage the cloud-specific set of `Deployments`, based on the selected `platform` in the `Infrastructure` resource.
+- `Deployments`, provisioned by the operator will run the pods on the `control-plane` nodes, and will tolerate `cloudprovider.kubernetes.io/uninitialized` and `node-role.kubernetes.io/master` taints.
+- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share a common approach with the bootstrap specific static-pod configuration.
+- Operator will own a `cluster-operator` resource, which will report the readiness of all workloads at post-install phase. The conditions reported will help other dependent components, such as [openshift-ingress](https://github.com/openshift/cluster-ingress-operator).
+- Operator will be built with a goal to achieve provider decoupling and potential switching between providers in the cluster, and will consider support to run multiple providers in a cluster simultaneously (long-term goal).
```

Comment:
Not sure what `switching between providers` mean here.

At first I understood you meant deploying on one platform, eg AWS, and later migrate your cluster to another platform such as OpenStack. It would be fantastic but this looks like science fiction to me and is way out of scope for this enhancement.

I assume you meant switching between different implementations of the same CCM?

Reply:
We could drop this, or reword, I've added a suggestion below which I think better explains how I had understood what this point was going for
```suggestion
- Operator will be built with consideration for potential future implementations of hybrid OpenShift clusters which support multiple cloud providers simultaneously (ie ensure resources created for each provider by the operator do not clash with resources for other providers)
```